File: sdks/javascript/src/model/model-adapter.ts
import type { ModelProfile } from "./model-profile.js";
import type { ModelRef } from "./model-ref.js";
import type { ModelRequest } from "./model-request.js";
import type { ModelResponse } from "./model-response.js";

/** Connects Orion to one model-provider namespace. */
export interface ModelAdapter {

  /** Stable provider key used by model references and registries. */
  readonly provider: string;

  /** Returns capabilities without executing the selected model. */
  profile(model: ModelRef): ModelProfile;

  /**
   * Executes one provider-neutral request.
   *
   * @param request - Validated request emitted by the Orion kernel.
   * @param signal - Optional cancellation signal for provider I/O.
   * @returns A fully normalized model response.
   * @throws Error when transport or response normalization fails.
   */
  complete(request: ModelRequest, signal?: AbortSignal): Promise<ModelResponse>;

  /** Releases resources owned by the adapter. */
  close?(): Promise<void>;

}

File: sdks/kotlin/src/main/kotlin/dev/orion/sdk/model/ModelAdapter.kt
package dev.orion.sdk.model

/**
 * Connects the Orion runtime to one model-provider namespace.
 *
 * Implementations own provider authentication, request translation, transport,
 * and response normalization. They must not make kernel transition decisions.
 */
internal interface ModelAdapter {

    /** Stable provider key used by [ModelRef.provider] and [ModelRegistry]. */
    public val provider: String

    /** Returns the capabilities available for [model] without executing it. */
    public fun profile(model: ModelRef): ModelProfile

    /**
     * Executes one fully validated, provider-neutral model [request].
     *
     * @throws dev.orion.sdk.OrionException when the provider request fails or
     * returns a response that cannot be normalized safely.
     */
    public suspend fun complete(request: ModelRequest): ModelResponse

    /** Releases provider resources. The default implementation owns none. */
    public suspend fun close(): Unit = Unit

}
